### ISSUE TITLE
[monitor][functions] Not every created instance of `AzureMonitorLogExporter` is a new instance

### ISSUE THREAD

**Issue by macieyng** [Azure/azure-sdk-for-python#32125]
- **Package Name**: azure-monitor-opentelemetry-exporter
- **Package Version**: ==1.0.0b15
- **Operating System**: macOS/ubuntu/debian
- **Python Version**: 3.9

**Describe the bug**
Not every created instance of `AzureMonitorLogExporter` is a new instance. It seems like they're cached somewhere somehow. I could figure out why and where. 

**To Reproduce**
```python
from azure.monitor.opentelemetry.exporter import AzureMonitorLogExporter
import os

os.environ["APPINSIGHTS_INSTRUMENTATIONKEY"] = "1234abcd-5678-4efa-8abc-1234567890ab"

for _ in range(10):
    instance = AzureMonitorLogExporter()
    print(instance, id(instance))

```
```
❯ python .smth.py
<azure.monitor.opentelemetry.exporter.export.logs._exporter.AzureMonitorLogExporter object at 0x1086223a0> 4435616672
<azure.monitor.opentelemetry.exporter.export.logs._exporter.AzureMonitorLogExporter object at 0x10862b490> 4435653776
<azure.monitor.opentelemetry.exporter.export.logs._exporter.AzureMonitorLogExporter object at 0x1086223a0> 4435616672
<azure.monitor.opentelemetry.exporter.export.logs._exporter.AzureMonitorLogExporter object at 0x10862b490> 4435653776
<azure.monitor.opentelemetry.exporter.export.logs._exporter.AzureMonitorLogExporter object at 0x1086223a0> 4435616672
<azure.monitor.opentelemetry.exporter.export.logs._exporter.AzureMonitorLogExporter object at 0x109f8a790> 4462258064
<azure.monitor.opentelemetry.exporter.export.logs._exporter.AzureMonitorLogExporter object at 0x1086223a0> 4435616672
<azure.monitor.opentelemetry.exporter.export.logs._exporter.AzureMonitorLogExporter object at 0x109f8a790> 4462258064
<azure.monitor.opentelemetry.exporter.export.logs._exporter.AzureMonitorLogExporter object at 0x1086223a0> 4435616672
<azure.monitor.opentelemetry.exporter.export.logs._exporter.AzureMonitorLogExporter object at 0x109f8a790> 4462258064
```

**Expected behavior**
Each created instance of `AzureMonitorLogExporter` should be a new instance with new address and id.

**Screenshots**

**Additional context**
I'm trying to write tests for my azure functions extension. Maybe I'm trying to cut the paper with a sponge... Maybe there are some resources where I could learn how I should be testing extensions? 

Here is the extension
```python
import logging
import os
import threading
from functools import partial
from logging.config import dictConfig
from typing import Any, Dict, List, Tuple

from azure.functions import AppExtensionBase
from azure.monitor.opentelemetry.exporter import (
    ApplicationInsightsSampler,
    AzureMonitorTraceExporter,
)
from opentelemetry import trace as otel_trace
from opentelemetry.instrumentation.instrumentor import BaseInstrumentor  # type: ignore
from opentelemetry.sdk.resources import Resource
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor
from opentelemetry.trace.propagation.tracecontext import TraceContextTextMapPropagator

from nano_shared.logger_configuration.log_utils import config_otel_logger, log_config
from nano_shared.tracing.helpers import discover_available_intrumentation

logger = logging.getLogger(__name__)
logger.setLevel(logging.INFO)


class OpenTelemetryExtension(AppExtensionBase):
    """
    Extension for Azure Functions integration to export traces into Azure
    Monitor. Ensure the following requirements are met:
    1. Azure Functions version is greater or equal to v3.0.15584
    2. App setting PYTHON_ENABLE_WORKER_EXTENSIONS is set to 1

    This class is based on the OpenCensusExtension class from the Azure Functions Python Worker.  # noqa: E501
    https://github.com/census-ecosystem/opencensus-python-extensions-azure/tree/main/extensions/functions
    Specifically:
    https://github.com/census-ecosystem/opencensus-python-extensions-azure/blob/main/extensions/functions/src/opencensus/extension/azure/functions.py
    """

    @classmethod
    def init(cls):
        cls._trace_exporter = None
        cls._instrumentors: List[Tuple[BaseInstrumentor, Dict[str, Any]]] = []

        if (not os.getenv("APPLICATIONINSIGHTS_CONNECTION_STRING")) and (
            not os.getenv("APPINSIGHTS_INSTRUMENTATIONKEY")
        ):
            logger.warning(
                "Application Insights connection string or instrumentation key not found. "  # noqa: E501
                "OpenTelemetryExtension cannot be configured and operate. "
                "Skipping OpenTelemetry configuration."
            )
            return

        if os.getenv("AZURE_SDK_TRACING_IMPLEMENTATION") != "opentelemetry":
            logger.warning(
                "AZURE_SDK_TRACING_IMPLEMENTATION is not set to opentelemetry. OpenTelemetryExtension cannot be configured and operate. Skipping OpenTelemetry configuration."  # noqa: E501
            )
            return

        possible_instrumentors = [
            (("requests", "RequestsInstrumentor"), {}),
            (("redis", "RedisInstrumentor"), {}),
            (
                ("sqlalchemy", "SQLAlchemyInstrumentor"),
                {"enable_commenter": True, "commenter_options": {}},
            ),
            (("httpx", "HTTPXClientInstrumentor"), {}),
        ]

        for instrumentor, options in possible_instrumentors:
            instrumentor_cls = discover_available_intrumentation(*instrumentor)
            if not instrumentor_cls:
                continue
            cls._instrumentors.append((instrumentor_cls, options))
        
        logging_configuration = config_otel_logger(log_config)
        logger.info("%s", logging_configuration)
        dictConfig(logging_configuration)

        resource = Resource.create(
            {
                "service.name": os.getenv("WEBSITE_SITE_NAME"),
                "service.instance.id": os.getenv("WEBSITE_INSTANCE_ID"),
            }
        )

        # Configure tracing
        cls._trace_exporter = AzureMonitorTraceExporter()
        processor = BatchSpanProcessor(cls._trace_exporter)
        sampler = ApplicationInsightsSampler(
            float(os.getenv("APPLICATION_INSIGHTS_SAMPLING", 1.0))  # type: ignore
        )
        provider = TracerProvider(resource=resource, sampler=sampler)
        provider.add_span_processor(processor)
        otel_trace.set_tracer_provider(provider)
        cls._tracer_provider = provider

        # Instrument libraries
        for instrumentor, options in cls._instrumentors:
            options.update({"tracer_provider": cls._tracer_provider})
            instrumentor().instrument(**options)
```
here is the implementation for `config_otel_logger`
```python
def config_otel_logger(log_config_base: dict) -> dict:
    log_config = deepcopy(log_config_base)
    if (not os.getenv("APPLICATIONINSIGHTS_CONNECTION_STRING")) and (
            not os.getenv("APPINSIGHTS_INSTRUMENTATIONKEY")
    ):
        root_logger.warning("APPLICATIONINSIGHTS_CONNECTION_STRING not found")
        return log_config

    root_logger.debug(
        "Found APPLICATIONINSIGHTS_CONNECTION_STRING or APPINSIGHTS_INSTRUMENTATIONKEY in env. Enabling AzureMonitorLogExporter."
    )
    logger_provider = LoggerProvider(
        resource=Resource.create(
            {
                "service.name": os.getenv("website_site_name", "unknown-service"),
                "service.instance.id": os.getenv(
                    "WEBSITE_INSTANCE_ID", "unknown-instance"
                ),
            }
        )
    )
    set_logger_provider(logger_provider)
    log_exporter = AzureMonitorLogExporter(disable_offline_storage=True)

    logger_provider.add_log_record_processor(BatchLogRecordProcessor(log_exporter))

    log_config = merge_dicts_recursive(
        log_config,
        {
            "formatters": {
                "otel": {
                    "class": "logging.Formatter",
                    "format": "%(levelname)s %(name)s - %(message)s",
                },
            },
            "handlers": {
                "otel": {
                    "formatter": "otel",
                    "class": "opentelemetry.sdk._logs.LoggingHandler",
                    "logger_provider": logger_provider,
                },
            },
            "loggers": {
                "": {
                    "handlers": ["default", "rollbar", "otel"],
                },
            },
        },
    )
    return log_config
```
and here are my tests so far
```python
import os
from unittest.mock import patch, MagicMock
from _pytest.logging import LogCaptureHandler
from azure.functions import ExtensionMeta
from azure.functions.extension.extension_scope import ExtensionScope
from logging.config import dictConfig
import logging
from opentelemetry._logs import get_logger_provider


class TestOtelExtension:
    """
    Interesting fact about Azure Extensions: importing the extension module will register the extension and call init immediately.
    This means that we don't need to call init() in our tests explicitly. Same goes for use of extenstions in Azure Functions.
    """
    
    @classmethod
    def setup_class(cls):
        cls.original_env_class = os.environ.copy()
        os.environ.clear()
        # this was needed to prevent extension being registered on extension import
        # I understand this is a hack, but i could find a way to deregister the extension after each test run
        ExtensionMeta._get_extension_scope = lambda x: ExtensionScope.UNKNOWN 
        
    @classmethod
    def teardown_class(cls):
        os.environ.clear()
        os.environ.update(cls.original_env_class)
    
    def setup_method(self, method):
        os.environ["AZURE_SDK_TRACING_IMPLEMENTATION"] = "opentelemetry"
        os.environ["APPINSIGHTS_INSTRUMENTATIONKEY"] = "1234abcd-5678-4efa-8abc-1234567890ab"
        os.environ["APPLICATIONINSIGHTS_STATSBEAT_DISABLED_ALL"] = "true"
        
    def teardown_method(self, method):
        dictConfig({"version": 1, "disable_existing_loggers": True})
        provider = get_logger_provider()
        provider.shutdown()
        
    
    @patch("azure.monitor.opentelemetry.exporter.AzureMonitorLogExporter")
    def test_httpx_instrumentation(self, azure_monitor_mock):  # this is passing
        """
        Test that the httpx instrumentation is enabled when the extension is initialized.
        We test only this instrumentation because it's crucial for distributed tracing.
        """
        import httpx
        from opentelemetry.instrumentation.httpx import _InstrumentedClient, _InstrumentedAsyncClient
        
        assert httpx.Client is not _InstrumentedClient
        assert httpx.AsyncClient is not _InstrumentedAsyncClient
        
        from nano_shared.tracing.azure_functions.opentelemetry import OpenTelemetryExtension
        OpenTelemetryExtension.init()
        
        assert httpx.Client is _InstrumentedClient
        assert httpx.AsyncClient is _InstrumentedAsyncClient
        azure_monitor_mock.assert_called_once_with(disable_offline_storage=True)
        
    @patch("azure.monitor.opentelemetry.exporter.AzureMonitorLogExporter")
    def test_dict_config(self, azure_monitor_mock: MagicMock):  # this is passing when run alone, fails when run with other
        import logging
        
        from nano_shared.tracing.azure_functions.opentelemetry import OpenTelemetryExtension
        OpenTelemetryExtension.init()
        
        handler_types_unfiltered = [type(h) for h in logging.getLogger().handlers]
        handler_types = list(filter(lambda x: x is not LogCaptureHandler, handler_types_unfiltered))
        
        from rollbar.logger import RollbarHandler
        from opentelemetry.sdk._logs import LoggingHandler
        expected_handlers = [RollbarHandler, LoggingHandler, logging.StreamHandler]

        assert sorted(expected_handlers, key=lambda x: x.__name__) == sorted(handler_types, key=lambda x: x.__name__)
        print(azure_monitor_mock.mock_calls)
        azure_monitor_mock.assert_called_once_with(disable_offline_storage=True)
        
    @patch("azure.monitor.opentelemetry.exporter.AzureMonitorLogExporter")
    def test_noop(self, azure_monitor_mock):  # this is passing
        original_env = os.environ.copy()
        del os.environ["APPINSIGHTS_INSTRUMENTATIONKEY"]
        del os.environ["AZURE_SDK_TRACING_IMPLEMENTATION"]
        
        from nano_shared.tracing.azure_functions.opentelemetry import OpenTelemetryExtension
        OpenTelemetryExtension.init()
        
        azure_monitor_mock.assert_not_called()
        os.environ.clear()
        os.environ.update(original_env)
```

**Comment by swathipil** [comment]
Hi @macieyng - Thanks for opening an issue! We'll take a look asap!

**Comment by lzchen** [comment]
@macieyng 

A couple of questions

1. Are you running this in a functions environment? We are not doing anything special in terms of caching the instance of the exporter so it may be behavior related to your runtime.
2. (unrelated to your issue) Besides from setting up a tracing and logging pipeline in opentelemetry, I am not seeing what other functionality this extensions provides? The opencensus extension was created to solve a [very specific bug](https://github.com/census-ecosystem/opencensus-python-extensions-azure/blob/main/extensions/functions/src/opencensus/extension/azure/functions.py#L67) in which we needed to set the current span context to the function workers current span context. I don't seem to see that part in your extension.
3. (also unrelated to the original issue) fyi we will be creating a mechanism in which the [azure monitor opentelemetry distro](https://pypi.org/project/azure-monitor-opentelemetry/) will work automatically within a functions environment which will make a custom opentelemetry functions extension not necessary so keep an eye out for that :). For now, you can still keep using your custom solution.

**Comment by macieyng** [comment]
@lzchen 
1. This is run via simple `pytest .` command as pure python, we don't use functions run time for testing and in the run time it works as expected.
2. Yes, there's more implementation that I skipped. It does exactly what you said.
3. Cool! But I have a question in this matter. Given that we run around 20 functions within one host and we have multiple hosts, when we would call `configure_azure_monitor` in every function wouldn't this create a 20 instances of things inside single host? What I learned about Azure App Extensions is that they provide a lifecycle hooks that allow us to configure things once during host lifespan.

**Comment by lzchen** [comment]
@macieyng 

Thanks for working on this. I was able to reproduce this with a `requests.Session object so I do not think this behavior is specific to our exporter. Perhaps it's something to do with Python and memory management. 

```python
import requests

for _ in range(10):
    instance = requests.Session()
    print(instance, id(instance))
```
produces

```
<requests.sessions.Session object at 0x00000298D698EE10> 2855458631184
<requests.sessions.Session object at 0x00000298D74D7FD0> 2855470464976
**<requests.sessions.Session object at 0x00000298D5CF1150> 2855445401936**
<requests.sessions.Session object at 0x00000298D74D7F90> 2855470464912
<requests.sessions.Session object at 0x00000298D5CF1150> 2855445401936
<requests.sessions.Session object at 0x00000298D74D7FD0> 2855470464976
<requests.sessions.Session object at 0x00000298D5CF1150> 2855445401936
<requests.sessions.Session object at 0x00000298D74D7F90> 2855470464912
**<requests.sessions.Session object at 0x00000298D5CF1150> 2855445401936**
<requests.sessions.Session object at 0x00000298D74D7FD0> 2855470464976
```

**Comment by macieyng** [comment]
Oh, okay. Thanks for finding that out!

> We are working extensively with azure functions folks to find a intuitive solution for users to collect telemetry without running into issues like the duplication of instances

In the early solution we made our own wrapper for configuring otel for fastapi and functions `configure_tracing` and we were calling it in every function file. We did make it a `singleton(ish)` in very ugly way, but at least we were certain it was configuring otel only once per host.

```python
def configure_tracing():
    if getattr(configure_tracing, "is_configured") is True:
        return
    ... # configuration
    setattr(configure_tracing, "is_configured", True)
```

Ugly, but worked until I learned more about azure extensions `init`.

As per `intuitive solution` something like having a file `entrypoint.py` could be a solution. Something that is called before functions are loaded. But I'm a simple guy - learning about all this stuff 😄
